### PR TITLE
feat: shelf and qtree plugin

### PR DESCRIPTION
**Pull request by Hardikl**
Shelf plugin is fully tested, Quota is in-progress. Logger level will be changed to debug later.

Compared rest metrics with zapi metrics:

Fan:
image

image

Sensor:
image

image

psu: [location is missing]
image

image

image

temperature:
image

image

voltage:
image

image

### CODE REVIEW

**Comment by rahulguptajss** [comment]
weird diff shows files from other commits? Any issues w.r.t to merging?

**Comment by Hardikl** [comment]
> weird diff shows files from other commits? Any issues w.r.t to merging?

yups, something went wrong in merging.

File: cmd/collectors/rest/plugins/shelf/shelf.go [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+					continue
+				}
+
+				if childObj := shelf.Get(attribute); childObj.Exists() {
+					if childObj.IsArray() {
+						for _, obj := range childObj.Array() {
+
+							// This is special condition, because child records can't be filterable in parent REST call
+							// frus type can be [module, psu] and we would only need psu for our use-case.
+							if attribute == "frus" && obj.Get("type").Exists() && obj.Get("type").String() == "module" {
```
Comment (rahulguptajss) [inline cmd/collectors/rest/plugins/shelf/shelf.go:231]: we do support filters in framework. You can try below.

api/storage/shelves?return_records=true&fields=*&query_fields=frus.type&query=psu

**Comment by Hardikl** [comment]
Due to some merge issue, closing this PR and created new one: #799

File: cmd/collectors/rest/plugins/shelf/shelf.go [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+		instanceKeys := exportOptions.NewChildS("instance_keys", "")
+		instanceKeys.NewChildS("", "shelf")
+		instanceKeys.NewChildS("", "channel")
+
+		// artificial metric for status of child object of shelf
+		my.data[attribute].NewMetricUint8("status")
+
+		for _, c := range childObj {
+
+			metricName, display := collector.ParseMetricName(c)
```
[the remark this answers is not in the document]
Comment (Hardikl) [reply cmd/collectors/rest/plugins/shelf/shelf.go:123]: There is circular dependency for using templating.go here

File: cmd/collectors/rest/plugins/shelf/shelf.go (hunk 1 above)
Comment (Hardikl) [reply cmd/collectors/rest/plugins/shelf/shelf.go:231]: filter and query can support main object but not able to handle the child array object.

**Pull request closed without merging**